### DISCUSSION TITLE
Is there a way to detect following data race code using golangci-lint or other linter??

### DISCUSSION THREAD

**Discussion by raeperd** [golangci/golangci-lint#5006] [Q&A] [answered]
Hi big fan of golangci-lint here! thanks a lot for saving me tons of times.

I have question about some hard to find data race error on code. 
[source is here](https://dave.cheney.net/2015/11/18/wednesday-pop-quiz-spot-the-race)

```go
package main

import (
        "fmt"
        "time"
)

type RPC struct {
        result int
        done   chan struct{}
}

func (rpc *RPC) compute() {
        time.Sleep(time.Second) // strenuous computation intensifies
        rpc.result = 42
        close(rpc.done)
}

func (RPC) version() int {
        return 1 // never going to need to change this
}

func main() {
        rpc := &RPC{done: make(chan struct{})}

        go rpc.compute()         // kick off computation in the background
        version := rpc.version() // grab some other information while we're waiting
        <-rpc.done               // wait for computation to finish
        result := rpc.result

        fmt.Printf("RPC computation complete, result: %d, version: %d\n", result, version)
}
```

By fixing `func (RPC) version()` to `func (RPC*) version` solves data race which is... very hard to noticed by human reader.
It would be great to catch this by linters

I tried to enable all linters in golangci-lint and many possible linters rules but none of linter catches this data race code. 
Am I missing something? or is there any linter that can catch?
Thanks for reading

**Comment by ldez** [comment]
Hello,

I'm not sure with a static analysis, it is possible to detect this race condition but it can be possible to detect receiver "inconsistency".

**Reply by raeperd** [reply]
Thanks for reply. is there any linter to check this inconsistency rule? I couldn't find any even outside of golangci-lint supported linters

**Reply by ldez** [reply, marked as answer]
Some IDEs handle this rule, but I think there is no existing linter on this topic.

**Reply by raeperd** [reply]
Thanks for reply 
I did extra research, I found another reference about this type of err

From [Go Wiki: Go Code Review Comments](https://go.dev/wiki/CodeReviewComments#receiver-type)
> Don’t mix receiver types. Choose either pointers or struct types for all available methods.
> Finally, when in doubt, use a pointer receiver.

Maybe I might try to build new linter to handle this kind of error..

**Comment by raeperd** [comment]
WIP in https://github.com/raeperd/recvcheck

**Reply by ldez** [reply]
I created a linter that aggregates the different needs around receivers.

- https://github.com/ldez/recv
- https://github.com/golangci/golangci-lint/pull/5013

Don't hesitate if you have feeback.

**Reply by raeperd** [reply]
@idenx 
Unfortunately, we did the almost same thing.. 🥲
I was actually looking forward to contributing to this issue, because golangci-lint is a tool I used a lot. And I was just preparing for PR.
If you don't mind, can I try merging the version of the code I made?

**Reply by ldez** [reply]
I'm sorry, I started working on that after my first answer.

My linter handles more things than yours:

- length of the receiver name
- consistency to the receiver type (pointer/non-pointer)
- consistency on the receiver names between the methods of a struct

And it uses a totally different approach.

Also, your linter forces the usage of a pointer receiver, which will lead to false positives because using a pointer is not always the right approach as suggested inside the [wiki](https://go.dev/wiki/CodeReviewComments#receiver-type).

> If you don't mind, can I try merging the version of the code I made?

Sadly, we cannot have several linters that handle the same rule.

You can open a PR but I will close it as a duplicate of "recv".

I will do the same with this PR https://github.com/golangci/golangci-lint/pull/5007

Sorry again, I should have talked about this sooner, but I didn't think you were going to make a linter.

**Reply by raeperd** [reply]
Since you're a maintainer, your opinion is more important of course, but honestly, it's true that I feel disrespected.  

I don't know man. I still let everyone know that I'm going to start working on it and that I'm in the process, and I didn't fly the PR as soon as I finished the code to do PR carefully in respect of goalngci-lint.

Whether it's your code or mine, I can add or improve functions together, so please give me a chance.

**Reply by ldez** [reply]
Be sure, there is no disrespect, I'm sorry once again.

> Whether it's your code or mine, I can add or improve functions together, so please give me a chance.

Of course, we can work together, there is no problem.  

Once again, sorry it was bad timing and bad communication on my side, I'm really sorry.

**Reply by raeperd** [reply]
Thank you for your apology. 
I appreciate your honesty and understand that miscommunication can happen. 
Will watch your PR and think about if there are any parts that I can contribute or improve.

**Reply by ldez** [reply]
I have an idea if you want: an improvement of "recv" can be to detect the usage of `sync.Mutex` as a struct field and in this specific context forces the usage of a pointer receiver.

I will not work on it, so feel free to do a PR on https://github.com/ldez/recv if you want.

**Reply by raeperd** [reply]
Thanks I'll try it tomorrow

**Reply by ldez** [reply]
We will change the plan: I will close my PR and we will use your linter.

But you will need to change your linter to do the same thing as mine: the report must be on the structure (not on the method) and no recommendation of a pointer receiver by default.

**Reply by raeperd** [reply]
@ldez 
Got it. I'll refer to your code to revise it and put you in the reviewer

**Reply by k713927** [reply]
> I have an idea if you want: an improvement of "recv" can be to detect the usage of `sync.Mutex` as a struct field and in this specific context forces the usage of a pointer receiver.
> 
> I will not work on it, so feel free to do a PR on https://github.com/ldez/recv if you want.

@raeperd  Hello, for this problem, have you updated it into the plug-in? If not, can I submit pr

**Reply by raeperd** [reply]
@k713927 Hello.
after my quick test and review, that problem is already solved by go vet: https://github.com/golang/tools/tree/master/go/analysis/passes/copylock
and this check is run by golangci-lint's govet linter
so i don't think your contribution is needed on this specific issue

**Reply by k713927** [reply]
> @k713927 Hello. after my quick test and review, that problem is already solved by go vet: https://github.com/golang/tools/tree/master/go/analysis/passes/copylock and this check is run by golangci-lint's govet linter so i don't think your contribution is needed on this specific issue

Oh, that's right. I hadn't noticed

**Comment by walidghallab** [comment]
Thanks for sharing the example. I have two comments about this:
1. Would you mind to share where is the race condition? You are saying that it is fixed by making Version ptr receiver, but never mentioned where is the race condition is and if it even exists.
2. The existence of this linter is good, but please note there are some cases where this is not recommended. For example "String()" method that implements fmt.Stringer, it is recommended in a lot of places to have it as a value receiver (regardless of what other methods uses) due to usability, otherwise if you implement it as pointer receiver anyone with a value to the struct (and not a pointer) won't have correct formatting, while value receiver works for both value and pointer.

**Reply by raeperd** [reply]
@walidghallab 
Hi sorry i'm late. I was busy on day job

1. 
Example code is completely executable, but data race can be found only when u add -race option
```
$ go build -race -o app - && •/app
WARNING: DATA RACE
Write at 0x00c000094020 by goroutine 6:
    main. (*RPC) . compute ()
        ~/recvcheck/main.go:15 +0x44
    main.main.gowrap1()
        ~/recvcheck/main.go:26 +0x20
        
Previous read at 0x00c000094020 by main goroutine:
    main-main ()
        ~/recvcheck/main. go:27 +0x120
        
Goroutine 6 (running) created at:
    main main ()
        ~/recvcheck/main.go: 26 +0x114
ニニニニニニニニニニニニニニニ
RPC computation complete, result: 42, version: 1
Found 1 data race(s)
```

2. 
About Stringer interface, if you mix receiver type then it also causes data race. 
I could say those recommendations are wrong in terms of concurrent safety.  
if you do not mix the type (using value type only), then you are fine

**Reply by LandonTClipp** [reply]
I'm confused at how this is a data race. The race checker seems to be keying off of the fact that the RPC struct itself is both read from and written to at the same time, but the RPC methods in question do not read or write the same attribute, so how can it be a data race? `Version()` is not even returning an attribute of the struct itself, it's returning an int literal.

It seems to me like Dave's example is overly contrived and doesn't present a real data race, and the race checker is giving a false positive. Someone tell me I'm wrong, I'm very confused 😅

**Reply by raeperd** [reply]
@LandonTClipp 
I'll help clarify the Go method and struct copy scenario related to potential data races. 

In Go, when a method is called on a struct, a copy of the entire struct is passed by default. This copying mechanism can introduce subtle data race conditions, especially when concurrent goroutines are involved. Even if the method doesn't directly use all struct fields, the entire struct is copied, which can create unexpected synchronization challenges.

For example, if you have a struct with some mutable fields and pass it to a method or goroutine, the copy might not prevent concurrent write operations, potentially leading to data races.

To prevent such issues, consider:
- Using pointers to structs
- Implementing proper synchronization (like mutexes)
- Using sync/atomic operations
- Leveraging Go's race detector during development

Key references:
- [Go Reddit Discussion on Struct Copying](https://www.reddit.com/r/golang/comments/10yrpic/why_not_mix_value_and_pointer_receivers/)
- [Go Race Detector Documentation](https://go.dev/doc/articles/race_detector)


and so, this new linter is needed 

https://github.com/raeperd/recvcheck

**Reply by karelbilek** [reply]
I still don't understand how is this a race and why does it matter. `version()` never reads any data from the struct, so it doesn't really matter that it might have incosistent `result` and `done`?

**Reply by LandonTClipp** [reply]
@raeperd sorry for the late reply, I wish Github was better about notifying me of replies!

Thank you for the explanation. I understand conceptually what you are saying in that making a copy of the struct creates the _potential_ for a data race, but the example we're given does not _actually_ demonstrate a data race because it does not show an attempt to read from and write to the same field of the struct. That's why I was asking for clarification because the example you posted is not a real data race, unless of course I am missing something obvious. Your example demonstrates a _potential_ data race but not a real one. 

I don't think I actually disagree or misunderstand anything of what you're saying here. My only confusion was that the example is not a real data race.
